go/libraries/doltcore/sqle/dropped_databases.go

```diff
@@ -0,0 +1,272 @@
+// dropped. The dolt_undrop() stored procedure is then able to restore them from this location.
+const droppedDatabaseDirectoryName = "dolt_dropped_databases"
+
+// droppedDatabaseManager is responsible for dropping databases and "undropping", or restoring, dropped databases. It
+// is given a Filesys where all database directories can be found. When dropping a database, instead of deleting the
+// database directory, it will move it to a new "dolt_dropped_databases" directory where databases can be restored.
+type droppedDatabaseManager struct {
+	fs filesys.Filesys
+}
+
+// newDroppedDatabaseManager creates a new droppedDatabaseManager instance using the specified |fs| as the location
+// where databases can be found. It will create a new "dolt_dropped_databases" directory at the root of |fs| where
+// dropped databases will be moved until they are permanently removed.
+func newDroppedDatabaseManager(fs filesys.Filesys) *droppedDatabaseManager {
+	return &droppedDatabaseManager{
+		fs: fs,
+	}
+}
+
+// DropDatabase will move the database directory for the database named |name| at the location |dropDbLoc| to the
+// dolt_dropped_database directory where it can later be "undropped" to restore it. If any problems are encountered
+// moving the database directory, an error is returned.
+func (dd *droppedDatabaseManager) DropDatabase(ctx *sql.Context, name string, dropDbLoc string) error {
+	rootDbLoc, err := dd.fs.Abs("")
+	if err != nil {
+		return err
+	}
+
+	isRootDatabase := false
+	// if the database is in the directory itself, we remove '.dolt' directory rather than
+	// the whole directory itself because it can have other databases that are nested.
+	if rootDbLoc == dropDbLoc {
+		doltDirExists, _ := dd.fs.Exists(dbfactory.DoltDir)
+		if !doltDirExists {
+			return sql.ErrDatabaseNotFound.New(name)
+		}
+		dropDbLoc = filepath.Join(dropDbLoc, dbfactory.DoltDir)
+		isRootDatabase = true
+	}
+
+	if err = dd.initializeDeletedDatabaseDirectory(); err != nil {
+		return fmt.Errorf("unable to drop database %s: %w", name, err)
+	}
+
+	// Move the dropped database to the Dolt deleted database directory so it can be restored if needed
+	_, file := filepath.Split(dropDbLoc)
+	var destinationDirectory string
+	if isRootDatabase {
+		// For a root database, first create the subdirectory before we copy over the .dolt directory
+		newSubdirectory := filepath.Join(droppedDatabaseDirectoryName, name)
+		if err := dd.fs.MkDirs(newSubdirectory); err != nil {
+			return err
+		}
+		destinationDirectory = filepath.Join(newSubdirectory, file)
+	} else {
+		destinationDirectory = filepath.Join(droppedDatabaseDirectoryName, file)
+	}
+
+	// Add the final directory segment and convert all hyphens to underscores in the database directory name
+	dir, file := filepath.Split(destinationDirectory)
+	if strings.Contains(file, "-") {
+		destinationDirectory = filepath.Join(dir, strings.ReplaceAll(file, "-", "_"))
```

Comment:
This means that the drop / undrop won't round-trip, which isn't great

There can also be other characters (anything non-alphanum) that get converted in a db name from its directory name, seems odd to special-case only the hyphen

Reply:
Maybe this is finally the time to stop messing with database names and just use the directory name verbatim (would be a breaking change, but I think it's a good idea)

Reply:
Regarding round-tripping... I think the most important requirement is that the database have the same logical/SQL name when it's restored. That would give us correct round-tripping from the SQL interface. I don't think round-tripping to the exact same physical directory name/location on disk is needed though, but it was something I was wondering about. Let me know if you see any potential issues with that. 

Good callout on non-alphanumeric chars also being translated. I'll find where we're handling that today for database names and see if I can pull something out to reuse. 

Stopping the logical DB renaming from physical directory name seems like a good idea. I think it's best done as a separate PR though – we should retain the old functionality for a while anyway, otherwise older DBs will not be able to load with newer Dolt versions (like we hit the other day).